rules/ast/ast.go

```diff
@@ -778,20 +793,81 @@ func (node *VectorArithExpr) Eval(timestamp clientmodel.Timestamp) Vector {
 			}
 		}
 		return result
+
 	} else if node.lhs.Type() == VectorType && node.rhs.Type() == VectorType {
 		lhs := node.lhs.(VectorNode).Eval(timestamp)
 		rhs := node.rhs.(VectorNode).Eval(timestamp)
-		for _, lhsSample := range lhs {
-			for _, rhsSample := range rhs {
-				if labelsEqual(lhsSample.Metric.Metric, rhsSample.Metric.Metric) {
-					value, keep := evalVectorBinop(node.opType, lhsSample.Value, rhsSample.Value)
-					if keep {
-						lhsSample.Value = value
-						if node.opType.shouldDropMetric() {
-							lhsSample.Metric.Delete(clientmodel.MetricNameLabel)
-						}
-						result = append(result, lhsSample)
+
+		// the control flow below handles one-to-one or many-to-one matching.
+		// for one-to-many swap sidedness and account for the swap when calculating
+		// values.
+		if node.matchType == MatchOTM {
+			lhs, rhs = rhs, lhs
+		}
+
+		// all samples from the rhs hashed by the matching label/values
+		rm := make(map[clientmodel.Fingerprint]*Sample)
+		// contains all fingerprints of rhs samples that have been used in an operation
+		// with a lhs sample. It is used to detect illegal many-to-one matchings.
+		added := make(map[clientmodel.Fingerprint]bool)
+
+		// add all rhs samples to a map so we can easily find matches later
+		for _, rs := range rhs {
+			fp := node.fingerprint(rs.Metric)
+			// the rhs is guaranteed to be the 'one' side. Having multiple samples
+			// with the same fingerprint means that the matching is many-to-many
+			// which is not supported
+			if rm[fp] != nil {
+				panic("many to many matching not allowed")
+			}
+			rm[fp] = rs
+		}
+
+		// for all lhs samples find a respective rhs sample and perform
+		// the binary operation
+		for _, ls := range lhs {
+			fp := node.fingerprint(ls.Metric)
+
+			rs, ok := rm[fp]
+			if ok {
+				if node.matchType == MatchOTO && added[fp] {
+					panic("many to one matching must be explicit")
+				}
+				// account for potentially swapped sidedness.
+				valLeft, valRight := ls.Value, rs.Value
+				if node.matchType == MatchOTM {
+					valLeft, valRight = valRight, valLeft
+				}
+				value, keep := evalVectorBinop(node.opType, valLeft, valRight)
+				if keep {
+					ns := &Sample{
+						Metric:    node.resultMetric(ls, rs),
+						Value:     value,
+						Timestamp: timestamp,
 					}
+					result = append(result, ns)
+					added[fp] = true
+					continue
+				}
+			}
+			// if we have an OR operation the union of both vectors is built and
```

Comment:
OR isn't a simple union, if there's a value for a labelset hash on the left you don't add in the one from the right